scipy/stats/_unuran/unuran_callback.h

```diff
@@ -0,0 +1,271 @@
+                                                                                       \
+    if (error) {                                                                       \
+        longjmp(err_callback->error_buf, 1);                                           \
+    }                                                                                  \
+                                                                                       \
+    return result
+
+
+/* A structure to represent a UNU.RAN Python callback */
+typedef struct unuran_callback {
+    /* A dictionary with required callbacks */
+    PyObject *callbacks;
+    /* Parameters (other than the quantile parameter) that the functions take */
+    PyObject *params;
+} unuran_callback_t;
+
+static ccallback_signature_t unuran_call_signatures[] = {
+    {NULL}
+};
+
+/**
+ * Initialize a UNU.RAN callback.
+ * 
+ * This function prepares a `ccallback_t` object containing the Python callbacks and
+ * their parameters. This helps obtain and call Python functions inside a C thunk in
+ * a thread-safe manner. Nonetheless, a callback needs to be aquired even when there
+ * are no Python callbacks because the error handler needs a `ccallback_t` object to
+ * jump to the Cython code where errors can be raised safely. In such a case, one can
+ * pass an empty dictionary in `fcn_dict` parameter.
+ * 
+ * Parameters
+ * ----------
+ * callback : ccallback_t *
+ *     Reference to an empty C callback
+ * unur_callback : unuran_callback_t *
+ *     Reference to an empty UNU.RAN callback
+ * fcn_dict : PyObject *
+ *     A Python dictionary containing the Python callbacks which are to be called in
+ *     the thunk. Can be an empty dict in case of no Python callbacks.
+ * extra_args : PyObject *
+ *     A tuple of extra arguments that need to be passed to the functions present in
+ *     the `fcn_dict` dictionary. If there are none, an empty tuple is expected.
+ * 
+ * Returns
+ * -------
+ * success : int
+ *     0 if success -1 otherwise
+ */
+int init_unuran_callback(ccallback_t *callback, unuran_callback_t *unur_callback,
+                         PyObject *fcn_dict, PyObject *extra_args)
+{
+    PyObject *dummy_fcn = NULL;
+    int ret;
+    int flags = CCALLBACK_OBTAIN;
+
+    callback->c_function = NULL;
+    callback->user_data = NULL;
+    callback->signature = NULL;
+
+    unur_callback->callbacks = fcn_dict;
+    Py_INCREF(fcn_dict);
+    unur_callback->params = extra_args;
+    Py_INCREF(extra_args);
+
+    /* The required callbacks are present in the `fcn_dict` dictionary. There can
+       be no callbacks for UNU.RAN yet we would need a C callback for error handler
+       to pass on the captured UNU.RAN errors to Python. Hence, we pack a dummy
+       function in the C callback object and use `unuran_callback_t` to obtain
+       required callbacks, if any. */
+    dummy_fcn = PyRun_String("lambda: None", Py_eval_input, PyEval_GetGlobals(), NULL);
```

Comment:
@rgommers @rkern Can one of you please review this bit of code here?

I currently take a dictionary of Python callbacks and put them in a `unuran_callback_t` struct. This is because multiple callbacks need to be called (e.g. "pdf", "dpdf", "cdf", etc) at each call to a UNU.RAN function and dictionaries help obtain them in the thunk easily without duplicating the thunks for each method. But it is also possible to have no Python callbacks, yet an error flag is needed to jump out of the C function and return the handle to Cython where the error can be thrown safely. So, a `ccallback_t` object needs to be initialized even when there are no Python callbacks. But it is not possible to initialize a `ccallback_t` object without passing it a Python callable or LowLevelCallable. So, to initialize the callback, I just create a dummy Python function using the very-high level CPython API. Is this OK to do so? or is there a way to avoid this workaround? (Sorry, for the very long and bad description. Hopefully, it makes some sense :) I have left some reviews to help understand what I am trying to do here.)

Reply:
This looks a bit unusual, and scary. I'm not sure I have enough context to understand the "multiple callbacks" issue, but for calling a Python function from C code you'd want `PyObject_CallFunction` (or some other `PyObject_Call*`) normally. If you search for `PyObject_Call` you'd find many examples in the SciPy code base.

For the multiple callbacks - maybe there can be a single private Python function/method that you call from C, and that function then calls pdf/cdf/etc,?

Reply:
> For the multiple callbacks - maybe there can be a single private Python function/method that you call from C, and that function then calls pdf/cdf/etc,?

I think this should work. I will try this out and see how it turns out. Still, the main problem is that I need a `ccallback_t` in the error handler for nonlocal return and, AFAIK, doing this is necessary to obtain one.

> This looks a bit unusual, and scary

I think `__minpack.h` does something similar to call jacobian functions. It has a `jac_callback_info_t` struct which is similar to the infrastucture I have here.

Reply:
> For the multiple callbacks - maybe there can be a single private Python function/method that you call from C, and that function then calls pdf/cdf/etc,?

Hi @rgommers! I have applied this suggestion now. I call a `self._callback_wrapper` in the thunk which in turn calls the given callback. Let me know what you think about it.

Reply:
This looks much nicer now, thanks!